src/server.rs

```diff
@@ -0,0 +1,395 @@
+use std::borrow::Cow;
+use std::io;
+
+use data_encoding::base64;
+use rand::distributions::IndependentSample;
+use rand::distributions::range::Range;
+use rand::os::OsRng;
+use rand::Rng;
+use ring::digest::Digest;
+
+use error::{Error, Kind, Field};
+use utils::find_proofs;
+use ::{NONCE_LENGTH, SHA256_LEN};
+
+/// Responds to client authentication challenges.
+/// The entrypoint for the SCRAM server side implementation
+pub struct ScramServer<P: AuthenticationProvider> {
+    /// The [authentication provider](trait.AuthenticationProvider.html) that will find passwords
+    /// and check authorization
+    provider: P
+
+}
+
+/// Contains information about stored passwords.  In particular, it stores the password that
+/// has been salted and hashed, the salt that was used, and the number of iterations of the hashing
+/// algorithm
+pub struct PasswordInfo {
+    hashed_password: Vec<u8>,
+    salt: Vec<u8>,
```

Comment:
I originally had this as a borrowed value, but switched it to owned after implementing `TestProvider`.  The issue with borrowing the password is that the object which owns it is the database query, which drops out of scope when the lookup method ends, as opposed to the database object itself, which lives longer.  It's possible with some memory trickery to switch ownership of the password to the database object, but that's hardly ideal.

As long as the password database is entirely stored in memory, then borrowed values work.  If there is any part which isn't, memory management becomes a nightmare, which is why I switched to owned, and took the small hit on copying.

However, there might be something obvious I'm missing here, so I'm perfectly happy to switch it to borrowed if that's the case.